src/v/security/gssapi_principal_mapper.cc

```diff
@@ -21,9 +21,57 @@
 #include <charconv>
 
 namespace security {
+static constexpr std::string_view gssapi_name_pattern
+  = R"(([^/@]*)(/([^/@]*))?@([^/@]*))";
 static constexpr std::string_view non_simple_pattern = R"([/@])";
 static constexpr std::string_view parameter_pattern = R"(([^$]*)(\$(\d*))?)";
 
+gssapi_name::gssapi_name(
+  std::string_view primary, std::string_view host_name, std::string_view realm)
+  : _primary(primary)
+  , _host_name(host_name)
+  , _realm(realm) {
+    if (_primary.empty()) {
+        throw std::invalid_argument("primary must be provided");
+    }
+}
+
+gssapi_name gssapi_name::parse(std::string_view principal_name) {
+    static thread_local const re2::RE2 gssapi_name_regex(
+      gssapi_name_pattern, re2::RE2::Quiet);
+    vassert(
+      gssapi_name_regex.ok(),
+      "Invalid name pattern: {}",
+      gssapi_name_regex.error());
+
+    re2::StringPiece primary, host_name, realm;
+
+    if (re2::RE2::FullMatch(
+          principal_name,
+          gssapi_name_regex,
+          &primary,
+          nullptr,
+          &host_name,
+          &realm)) {
+        return gssapi_name(
+          ss::sstring(primary), ss::sstring(host_name), ss::sstring(realm));
```

Comment:
nit: you could return `{ ... }` here instead of repeating the name of the function's return type